crates/astria-sequencer/src/ibc/ics20_transfer.rs

```diff
@@ -288,290 +353,280 @@ impl AppHandlerExecute for Ics20Transfer {
-            .is_ok()
-    {
-        execute_withdrawal_refund_to_rollup(state, packet_data)
-            .await
-            .context("failed to execute rollup withdrawal refund")?;
-        return Ok(());
-    }
-
-    let packet_amount: u128 = packet_data
+    let amount: u128 = packet_data
         .amount
         .parse()
         .context("failed to parse packet data amount to u128")?;
-    let mut denom_trace = {
-        let denom = packet_data
-            .denom
-            .parse::<Denom>()
-            .context("failed parsing denom in packet data as Denom")?;
-        // convert denomination if it's prefixed with `ibc/`
-        // note: this denomination might have a prefix, but it wasn't prefixed by us right now.
-        convert_denomination_if_ibc_prefixed(state, denom)
-            .await
-            .context("failed to convert denomination if ibc/ prefixed")?
-    };
 
-    // the IBC packet should have the address as a bech32 string
-    let recipient = if is_refund {
-        packet_data.sender.clone()
-    } else {
-        packet_data.receiver
-    };
-    let recipient = recipient.parse().context("invalid recipient address")?;
+    let recipient = packet_data
+        .receiver
+        .parse()
+        .context("invalid recipient address")?;
 
-    let is_prefixed = denom_trace.starts_with_str(&format!("{source_port}/{source_channel}"));
-    let is_source = if is_refund {
-        // we are the source if the denom is not prefixed by source_port/source_channel
-        !is_prefixed
+    let mut asset = parse_asset(&state, &packet_data.denom)
+        .await
+        .with_context(|| {
+            format!(
+                "failed reading asset `{}` from packet data",
+                packet_data.denom
+            )
+        })?;
+
+    let is_source = is_transfer_source_zone(&asset, &packet.port_on_a, &packet.chan_on_a);
+    if is_source {
+        asset.pop_leading_port_and_channel();
     } else {
-        // we are the source if the denom is prefixed by source_port/source_channel
-        is_prefixed
-    };
+        // TODO(janis): we should provide a method on `denom::TracePrefixed` to directly
+        // prefix it with a (port, channel) pair.
+        asset = format!(
+            "{destination_port}/{destination_channel}/{asset}",
+            destination_port = &packet.port_on_b,
+            destination_channel = &packet.chan_on_b,
+        )
+        .parse()
+        .expect(
+            "dest port and channel are valid prefix segments, so this concatenation must be a \
+             valid denom",
+        );
+    }
 
-    // prefix the denomination with the destination port and channel if not a refund
-    let trace_with_dest =
-        prepend_denom_if_not_refund(&denom_trace, dest_port, dest_channel, is_refund);
-
-    // check if this is a transfer to a bridge account and
-    // execute relevant state changes if it is
-    execute_ics20_transfer_bridge_lock(
```

Comment:
I believe this function was incorrect: its argument `trace_with_dest = dest_port/dest_channel/source_port/source_channel/asset` was always constructed irrespective of whether sequencer was a source or a sink zone, but the construction is only correct if sequencer is a sink!

If sequencer is a source, then `dest_port/dest_channel` must *not* be prepended, and instead `source_port/soure_channel` removed, leaving just `asset`.